### PR TITLE
Chapter 10: Account Activation and Password Reset

### PR DESCRIPTION
**Pull request by JOlsen-Neslo**

### CODE REVIEW

File: app/controllers/password_resets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  def password_reset_params
+    params.require(:password_reset).permit(:email)
+  end
+
+  def get_user
+    @user = User.find_by(email: params[:email])
+  end
+
+  # Confirms a valid user.
+  def valid_user
```
Comment (bradleymarques) [inline app/controllers/password_resets_controller.rb]: What about `validate_user`?

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -24,4 +22,15 @@ def destroy
   def session_params
     params.require(:session).permit(:email, :password, :remember_me)
   end
+
+  def check_if_activated(user)
```
Comment (bradleymarques) [inline app/controllers/sessions_controller.rb]: This method seems to be doing more than checking if the user is activated.  What about renaming the method to capture the logic and consequences?

File: app/models/user.rb [hunk 3]
```diff
@@ -29,4 +30,45 @@ def remember
...
     update_attribute(:remember_digest, nil)
   end
+
+  # Activates an account.
+  def activate
+    update_attribute(:activated, true)
+    update_attribute(:activated_at, Time.zone.now)
+  end
+
+  # Sends activation email.
```
Comment (bradleymarques) [inline app/models/user.rb]: Pointless comment

File: app/models/user.rb [hunk 4]
```diff
@@ -29,4 +30,45 @@ def remember
   def forget
     update_attribute(:remember_digest, nil)
   end
+
+  # Activates an account.
+  def activate
```
Comment (bradleymarques) [inline app/models/user.rb]:
```suggestion
  def activate_account
```

And kill the comment

File: app/models/user.rb [hunk 5]
```diff
@@ -29,4 +30,45 @@ def remember
...
+  end
+
+  # Sets the password reset attributes.
+  def create_reset_digest
+    self.reset_token = new_token
+    update_attribute(:reset_digest,  digest(reset_token))
+    update_attribute(:reset_sent_at, Time.zone.now)
+  end
+
+  # Sends password reset email.
```
Comment (bradleymarques) [inline app/models/user.rb]: Pointless comment

File: app/models/user.rb [hunk 6]
```diff
@@ -29,4 +30,45 @@ def remember
...
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end
+
+  # Returns true if a password reset has expired.
```
Comment (bradleymarques) [inline app/models/user.rb]: Pointless comment

File: app/views/user_mailer/account_activation.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<h1>Sample App</h1>
```
Comment (bradleymarques) [inline app/views/user_mailer/account_activation.html.erb]: I18n

File: app/views/user_mailer/account_activation.html.erb [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+<p>Hi <%= @user.name %>,</p>
+
+<p>
+    Welcome to the Sample App! Click on the link below to activate your account:
```
Comment (bradleymarques) [inline app/views/user_mailer/account_activation.html.erb]: I18n

File: app/views/user_mailer/account_activation.html.erb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<h1>Sample App</h1>
+
+<p>Hi <%= @user.name %>,</p>
```
Comment (bradleymarques) [inline app/views/user_mailer/account_activation.html.erb]: I18n (you can use interpolation to pass in the name)

https://stackoverflow.com/questions/6536304/how-can-i-perform-interpolation-in-i18n-locales

File: app/views/user_mailer/account_activation.text.erb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+Hi <%= @user.name %>,
```
Comment (bradleymarques) [inline app/views/user_mailer/account_activation.text.erb]: I18n for all hard-coded strings

File: app/models/concerns/authentication_concern.rb [hunk 11]
```diff
@@ -16,9 +16,9 @@ def determine_cost
     end
 
     # Returns true if the given token matches the digest.
-    def authenticated?(remember_token, remember_digest = nil)
-      return false if remember_digest.nil?
-      BCrypt::Password.new(remember_digest).is_password?(remember_token)
+    def authenticated?(token, digest = nil)
+      return false if digest.nil?
+      BCrypt::Password.new(digest).is_password?(token)
```
Comment (c13e) [inline app/models/concerns/authentication_concern.rb]: You can also use the `&&` operator which 'short-circuits' (returns left if true, else returns right)
```ruby
digest.present? && BCrypt::Password.new(digest).is_password?(token)
```

File: app/views/password_resets/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+            <%= f.label :password %>
+            <%= f.password_field :password, class: 'form-control' %>
+
+            <%= f.label :password_confirmation, I18n.t 'resets.edit.confirmation' %>
+            <%= f.password_field :password_confirmation, class: 'form-control' %>
+
+            <%= f.submit I18n.t 'resets.edit.submit', class: "btn btn-primary" %>
+        <% end %>
+    </div>
+</div>
```
Comment (c13e) [inline app/views/password_resets/edit.html.erb]: Add a new line here :)

File: app/views/password_resets/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+<div class="row">
+    <div class="col-md-6 col-md-offset-3">
+        <%= form_for(:password_reset, url: password_resets_path) do |f| %>
+            <%= f.label :email %>
+            <%= f.email_field :email, class: 'form-control' %>
+
+            <%= f.submit I18n.t 'resets.submit', class: "btn btn-primary" %>
+        <% end %>
+    </div>
+</div>
```
Comment (c13e) [inline app/views/password_resets/new.html.erb]: Ditto

**Pull request merged**